File: cs/days/day06.cs
namespace Shunty.AoC;

public class Day06 : AocDaySolution
{
    public async Task Run()
    {
        var data = await AocUtils.GetDayText(6);
        var p1 = ScanInput(data.AsSpan(), 4);
        var p2 = ScanInput(data.AsSpan(), 14);

        Console.WriteLine("Day 6");
        Console.WriteLine($"  Part 1: {p1}");
        Console.WriteLine($"  Part 2: {p2}");
    }

    private int ScanInput(ReadOnlySpan<char> input, int packetLength)
    {
        var i = 0;
        while (i < input.Length)
        {
            var (success, idx) = ScanPacket(input.Slice(i, packetLength));
            if (success)
                return i + packetLength;
            i += idx;
        }
        throw new Exception("Unable to find an appropriate packet");
    }

    private (bool success, int dupIndex) ScanPacket(ReadOnlySpan<char> pkt)
    {
        var i = 1;
        foreach (var ch in pkt)
        {
            if (pkt.Slice(i).Contains(ch))
            {
                return (false, i);
            }
            i++;
        }
        return (true, 0);
    }
}

File: py/day06.py
import os

# https://adventofcode.com/2022/day/6 - Day 6: Tuning Trouble

def check_packet(pkt: str) -> tuple[bool, int]:
    for i in range(len(pkt)):
        if (pkt[i] in pkt[i+1:]):
            return (False, i+1)
    return (True, 0)

def scan_input(input: str, packet_len: int) -> int:
    i = 0
    while i < len(input):
        (success, index) = check_packet(input[i:i+packet_len])
        if success:
            return i + packet_len
        i += index
    raise IndexError("Unable to find an appropriate packet")

## main

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day06-input", "r") as f:
    data: str = f.read()

print("Part 1: ", scan_input(data, 4))
print("Part 2: ", scan_input(data, 14))
